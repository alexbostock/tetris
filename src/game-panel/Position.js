// @flow

class Position {
  x: number;
  y: number;

  constructor(x: number, y: number) {
    this.x = x;
    this.y = y;
  }

  downOne() {
    return new Position(this.x, this.y + 1);
  }

  upOne() {
    return new Position(this.x, this.y - 1);
  }

  rightOne() {
    return new Position(this.x + 1, this.y);
  }

  leftOne() {
    return new Position(this.x - 1, this.y);
  }

  add(p: Position) {
    return new Position(this.x + p.x, this.y + p.y);
  }

  equals(other: Position) {
    return this.x === other.x && this.y === other.y;
  }

  hashCode() {
    return this.x + this.y | 0;
  }
}

export default Position;