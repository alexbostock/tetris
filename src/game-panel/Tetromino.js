// @flow

import { Set } from 'immutable';

import Position from './Position';

type TetrominoType =
  | 'I'
  | 'J'
  | 'L'
  | 'O'
  | 'S'
  | 'T'
  | 'Z';

class Tetromino {
  type: TetrominoType;
  pos: Position;
  // Use Super Rotation System (https://tetris.fandom.com/wiki/SRS).
  orientation: number;

  constructor(type: TetrominoType, pos: Position = new Position(4, 0)) {
    this.type = type;
    this.pos = pos;
    this.orientation = 0;
  }

  occupiedCells() {
    const cells = cellsOrientation0(this.type, this.pos);

    switch (this.type) {
      case 'I': return rotateI(cells, this.pos, this.orientation);
      case 'O': return cells;
      default: return rotate(cells, this.pos.downOne(), this.orientation);
    }
  }

  rotateLeft() {
    const t = new Tetromino(this.type, this.pos);
    t.orientation = (this.orientation + 3) % 4;
    return t;
  }

  rotateRight() {
    const t = new Tetromino(this.type, this.pos);
    t.orientation = (this.orientation + 1) % 4;
    return t;
  }
}

function cellsOrientation0(type: TetrominoType, centre: Position) {
  switch (type) {
    case 'I':
      return Set([
        centre,
        centre.leftOne(),
        centre.rightOne(),
        centre.rightOne().rightOne(),
      ]);
    case 'J':
      return Set([
        centre.leftOne(),
        centre.downOne().leftOne(),
        centre.downOne(),
        centre.downOne().rightOne(),
      ]);
    case 'L':
      return Set([
        centre.rightOne(),
        centre.downOne().leftOne(),
        centre.downOne(),
        centre.downOne().rightOne(),
      ]);
    case 'O':
      return Set([
        centre,
        centre.downOne(),
        centre.rightOne(),
        centre.downOne().rightOne(),
      ]);
    case 'S':
      return Set([
        centre,
        centre.rightOne(),
        centre.downOne(),
        centre.downOne().leftOne(),
      ]);
    case 'T':
      return Set([
        centre,
        centre.downOne().leftOne(),
        centre.downOne(),
        centre.downOne().rightOne(),
      ]);
    case 'Z':
      return Set([
        centre,
        centre.leftOne(),
        centre.downOne(),
        centre.downOne().rightOne(),
      ]);
    default:
      console.error(`Unexpected tetromino type: ${type}`);
      return Set();
  }
}

function rotate(points: Set<Position>, centre, numRotations) {
  let ps = points.map(p => new Position(p.x - centre.x, p.y - centre.y));

  for (let i = 0; i < numRotations; i++) {
    ps = ps.map(p => new Position(-p.y, p.x));
  }

  return ps.map(p => new Position(p.x + centre.x, p.y + centre.y));
}

function rotateI(points: Set<Position>, centre, orientation) {
  switch (orientation) {
    case 0: return points;
    case 2: return points.map(p => new Position(p.x, p.y + 1));
    case 1:
      centre = new Position(centre.x + 1, centre.y);
      // Fallthrough
    case 3:
      return Set([
        centre,
        centre.upOne(),
        centre.downOne(),
        centre.downOne().downOne(),
      ]);
    default:
      console.error(`Unexpected tetromino orientation: ${orientation}`);
        return Set();
  }
}

class Generator {
  // Use Random Generator (https://tetris.fandom.com/wiki/Random_Generator).
  
  bag: Array<TetrominoType> = [];

  next = () => {
    if (this.bag.length === 0) {
      this.bag = ['I', 'J', 'L', 'O', 'S', 'T', 'Z'];
    }

    const index = Math.floor(Math.random() * this.bag.length);
    return new Tetromino(this.bag.splice(index, 1)[0]);
  }
}

export const tetrominoGenerator = new Generator();

export default Tetromino;