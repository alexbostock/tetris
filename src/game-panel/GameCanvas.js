// @flow

import React from 'react';
import { Set } from 'immutable';

import Position from './Position';
import Tetromino from './Tetromino';

type Props = {
  tetromino: Tetromino,
  staticBlocks: Set<Position>,
  addStaticBlock: Position => void,
};

function GameCanvas(props: Props) {
  const height = 20;
  const width = 10;

  return (
    <div
      id="gameCanvas"
      style={{
        display: 'grid',
        gridTemplateColumns: width,
        gridTemplateRows: height,
      }}
    >
      {props.tetromino.occupiedCells().map(pos => Mino(pos, 'red'))}
      {props.staticBlocks.map(pos => Mino(pos, 'blue'))}
    </div>
  );
}

function Mino(pos: Position, colour: string) {
  const style = {
    gridColumn: pos.x,
    gridRow: pos.y,
    color: colour,
  };

  return <div key={pos.x + ',' + pos.y} className="mino" style={style}></div>;
}

export default GameCanvas;