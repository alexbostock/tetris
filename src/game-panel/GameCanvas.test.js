import React from 'react';
import renderer from 'react-test-renderer';
import { Set } from 'immutable';

import Position from './Position';
import Tetromino from './Tetromino';
import GameCanvas from './GameCanvas';

test('correctly renders minos described by props', () => {
  const tet = new Tetromino('I', new Position(4, 0));
  const staticBlocks = Set([
    new Position(3, 10),
    new Position(4, 10),
    new Position(5, 10),
  ]);

  const canvas = renderer.create(
    <GameCanvas
      tetromino={tet}
      staticBlocks={staticBlocks}
      addStaticBlock={x => x}
    />
  );

  expect(canvas.toJSON()).toMatchSnapshot();
});
