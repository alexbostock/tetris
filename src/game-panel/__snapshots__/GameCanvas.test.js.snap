// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`correctly renders minos described by props 1`] = `
<div
  id="gameCanvas"
  style={
    Object {
      "gridTemplateColumns": "repeat(10, 1fr)",
      "gridTemplateRows": "repeat(20, 1fr)",
      "height": "60vh",
      "width": "30vh",
    }
  }
>
  <div
    className="mino"
    style={
      Object {
        "backgroundColor": "red",
        "gridColumn": 5,
        "gridRow": 1,
        "opacity": 1,
      }
    }
  />
  <div
    className="mino"
    style={
      Object {
        "backgroundColor": "red",
        "gridColumn": 4,
        "gridRow": 1,
        "opacity": 1,
      }
    }
  />
  <div
    className="mino"
    style={
      Object {
        "backgroundColor": "red",
        "gridColumn": 6,
        "gridRow": 1,
        "opacity": 1,
      }
    }
  />
  <div
    className="mino"
    style={
      Object {
        "backgroundColor": "red",
        "gridColumn": 7,
        "gridRow": 1,
        "opacity": 1,
      }
    }
  />
  <div
    className="mino"
    style={
      Object {
        "backgroundColor": "brown",
        "gridColumn": 4,
        "gridRow": 11,
        "opacity": 1,
      }
    }
  />
  <div
    className="mino"
    style={
      Object {
        "backgroundColor": "brown",
        "gridColumn": 5,
        "gridRow": 11,
        "opacity": 1,
      }
    }
  />
  <div
    className="mino"
    style={
      Object {
        "backgroundColor": "brown",
        "gridColumn": 6,
        "gridRow": 11,
        "opacity": 1,
      }
    }
  />
  <div
    id="startGameMenu"
  >
    <button
      type="button"
    >
      START GAME
    </button>
    <button
      type="button"
    >
      SHOW LEADERBOARD
    </button>
  </div>
</div>
`;

exports[`correctly renders minos described by props 2`] = `
<div
  id="gameCanvas"
  style={
    Object {
      "gridTemplateColumns": "repeat(10, 1fr)",
      "gridTemplateRows": "repeat(20, 1fr)",
      "height": "60vh",
      "width": "30vh",
    }
  }
>
  <div
    className="mino"
    style={
      Object {
        "backgroundColor": "red",
        "gridColumn": 5,
        "gridRow": 1,
        "opacity": 1,
      }
    }
  />
  <div
    className="mino"
    style={
      Object {
        "backgroundColor": "red",
        "gridColumn": 4,
        "gridRow": 1,
        "opacity": 1,
      }
    }
  />
  <div
    className="mino"
    style={
      Object {
        "backgroundColor": "red",
        "gridColumn": 6,
        "gridRow": 1,
        "opacity": 1,
      }
    }
  />
  <div
    className="mino"
    style={
      Object {
        "backgroundColor": "red",
        "gridColumn": 7,
        "gridRow": 1,
        "opacity": 1,
      }
    }
  />
  <div
    className="mino"
    style={
      Object {
        "backgroundColor": "brown",
        "gridColumn": 4,
        "gridRow": 11,
        "opacity": 1,
      }
    }
  />
  <div
    className="mino"
    style={
      Object {
        "backgroundColor": "brown",
        "gridColumn": 5,
        "gridRow": 11,
        "opacity": 1,
      }
    }
  />
  <div
    className="mino"
    style={
      Object {
        "backgroundColor": "brown",
        "gridColumn": 6,
        "gridRow": 11,
        "opacity": 1,
      }
    }
  />
  <div
    id="startGameMenu"
  >
    <h2>
      GAME OVER
    </h2>
    <button
      type="button"
    >
      PLAY AGAIN
    </button>
    <form>
      <h3>
        SUBMIT TO LEADERBOARD
      </h3>
      <label>
        NICKNAME:
        <input
          onChange={[Function]}
          value=""
        />
        <button
          disabled={true}
          onClick={[Function]}
        >
          SUBMIT
        </button>
      </label>
      <p>
        SAVING
      </p>
    </form>
    <button
      type="button"
    >
      SHOW LEADERBOARD
    </button>
  </div>
</div>
`;

exports[`correctly renders minos described by props 3`] = `
<div
  id="gameCanvas"
  style={
    Object {
      "gridTemplateColumns": "repeat(10, 1fr)",
      "gridTemplateRows": "repeat(20, 1fr)",
      "height": "60vh",
      "width": "30vh",
    }
  }
>
  <div
    className="mino"
    style={
      Object {
        "backgroundColor": "red",
        "gridColumn": 5,
        "gridRow": 1,
        "opacity": 1,
      }
    }
  />
  <div
    className="mino"
    style={
      Object {
        "backgroundColor": "red",
        "gridColumn": 4,
        "gridRow": 1,
        "opacity": 1,
      }
    }
  />
  <div
    className="mino"
    style={
      Object {
        "backgroundColor": "red",
        "gridColumn": 6,
        "gridRow": 1,
        "opacity": 1,
      }
    }
  />
  <div
    className="mino"
    style={
      Object {
        "backgroundColor": "red",
        "gridColumn": 7,
        "gridRow": 1,
        "opacity": 1,
      }
    }
  />
  <div
    className="mino"
    style={
      Object {
        "backgroundColor": "brown",
        "gridColumn": 4,
        "gridRow": 11,
        "opacity": 1,
      }
    }
  />
  <div
    className="mino"
    style={
      Object {
        "backgroundColor": "brown",
        "gridColumn": 5,
        "gridRow": 11,
        "opacity": 1,
      }
    }
  />
  <div
    className="mino"
    style={
      Object {
        "backgroundColor": "brown",
        "gridColumn": 6,
        "gridRow": 11,
        "opacity": 1,
      }
    }
  />
</div>
`;
